src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -202,57 +202,132 @@ public void lazy_mapping() {
         }
 
         public List<R> force() {
-            List<R> result = new ArrayList<>();
-            for (T element : list) {
-                result.addAll(function.apply(element));
-            }
-            return result;
+            return traversable.force();
         }
 
-        public LazyFlatMapHelper<T, R> filter (Predicate<T> predicate){
-            return new LazyFlatMapHelper<>(
-                list,
-                    element -> predicate.test(element) ?
-                            function.apply(element) :
-                            Collections.emptyList()
-            );
+        public LazyFlatMapHelper<T, R> filter(Predicate<R> predicate) {
+            return new LazyFlatMapHelper<>(list, traversable.filter(predicate));
         }
 
-        public <R2> LazyFlatMapHelper<T, R2> map(Function<R, R2> f) {
-            final Function<R, List<R2>> listFunction = rR2TorListR2(f);
-            return flatMap(listFunction);
+        public <R2> LazyFlatMapHelper<T, R2> map(Function<R, R2> mapper) {
+            return new LazyFlatMapHelper<>(list, traversable.map(mapper));
         }
 
-        // (R -> R2) -> (R -> [R2])
-        private <R2> Function<R, List<R2>> rR2TorListR2(Function<R, R2> mapper) {
-            return element -> Collections.singletonList(mapper.apply(element));
+        public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> f) {
+            return new LazyFlatMapHelper<>(list, traversable.flatMap(f));
         }
+    }
 
-        // TODO *
-        public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> f) {
-            return new LazyFlatMapHelper<>(
-                    list,
+    private interface Traversable<T> {
+        void forEach(Consumer<T> consumer);
+
+        static <T> Traversable<T> from(final List<T> list) {
+            return list::forEach;
+        }
+
+        default Traversable<T> filter(final Predicate<T> predicate) {
+            final Traversable<T> self = this;
+            return consumer -> self.forEach(
                     element -> {
-                        List<R2> results = new ArrayList<>();
-                        function.apply(element).forEach(result -> results.addAll(f.apply(result)));
-                        return results;
-                    });
+                        if (predicate.test(element)) consumer.accept(element);
+                    }
+            );
+        }
+
+        default <R> Traversable<R> map(final Function<T, R> mapper) {
+            final Traversable<T> self = this;
+            return consumer -> self.forEach(
+                    element -> consumer.accept(mapper.apply(element))
+            );
         }
+
+        default <R> Traversable<R> flatMap(final Function<T, List<R>> function) {
+            final Traversable<T> self = this;
+            return consumer -> self.forEach(
+                    collection -> function.apply(collection).forEach(consumer)

```

Comment:
It equals the type of the objects,that are stored in the given list.
Renamed the variable name